### ISSUE TITLE
Make tyk-python independent from the system Python version

### ISSUE THREAD

**Issue by excieve** [TykTechnologies/tyk#1283]
**Do you want to request a *feature* or report a *bug*?**
Feature

**What is the current behavior?**
We're dynamically linking against `libpython3`, however our packages are built for Python 3.4, which is an issue outside of a vetted environment (e.g. docker). Even when targeting a specific distro version, system can upgrade a Python minor, which may break ABI (current Ubuntu LTS has 3.5). This implies laying additional burden on infrastructure of the users who use raw packages and prevents us from easily upgrading even our own docker base images.

**What is the expected behavior?**
Statically link Python runtime and bundle its standard library along with our required dependencies so that we wouldn't depend on a specific version of Python being present on the system.

The proposed solution will make our packages fully self-contained in terms of dependencies, make deployments more simple for the users and let us easily manage it without worrying for compatibility. The downsides are:
 1) increasing the size of the package and executable (but it's not by a lot so I wouldn't call it an issue)
 2) having to find a way for plugin developers to add 3rd-party dependencies as it wouldn't be possible to just use system-wide pip for it

The latter could be solved in a user-friendly way with e.g. `tyk-cli` reading the standard `requirements.txt` and installing in the bundle or maybe creating a virtual env for it, either way it's solvable.

Other possible solutions to the issue are:
 1) Setting `Py_LIMITED_API` preprocessor macro to only use a stable subset of Python API that never breaks ABI for Python 3.2+. Unfortunately this doesn't currently work (tested it) due to our reliance on Cython, which does not support it and [doesn't seem to be keen on it any time soon](https://mail.python.org/pipermail/cython-devel/2011-February/000108.html). Is there a way for us to more or less painlessly drop Cython?
 2) Compiling binaries for every Python 3.4+ minor version. This would work but it's quite ugly IMO and would require us to add more binaries with every minor release of Python.

Other points of interest:
 1) Linking `libpython` statically is actually a preferred mode for the interpreter as it can win about 9% in performance according to some [benchmark](https://gist.github.com/methane/4681131).
 2) Although by default cgo links dynamically, it's capable of static linking as well so it doesn't really require any magic.
 3) Self-contained package means we can easily support any Linux distro and any release. It even makes building `tyk-python` for archs other than `x86-64` much more viable.
 4) The Python stdlib and our dependencies (protobuf, grpcio) can be bundled in a zip file to decrease the size and number of files in the package. Python's loader can work with it automagically. It's the same thing Sublime Text and Dropbox does with their Python runtime.
 5) Python's license, while GPL-compatible, is very permissive for derivatives so it's not a problem at all to include and redistribute like this. Some dependencies of Python might require their licenses copied verbatim somewhere though.

**Comment by excieve** [comment]
@matiasinsaurralde FYI, what do you think?

**Comment by matiasinsaurralde** [comment]
We discussed this last year, would you be able to help on creating a static build for this?
I agree with all the points, I think that we should also look forward to use the minimal API, our glue code is very small so switching to the minimal API shouldn't be an issue, I can take that myself.

**Comment by matiasinsaurralde** [comment]
Just finished testing our current code with the minimal API under Linux, it doesn't break so we won't require any major code changes in the glue code.
About Cython: we used it because it was an easier way of adding `cffi` calls.

**Comment by buger** [comment]
@excieve if you can make this work it will be great!

**Comment by excieve** [comment]
@matiasinsaurralde Cython-generated code is what currently breaks it for me as it uses C API outside of the stable subset a lot. If there's a way to sanely drop it and use limited API all around — it should probably be better than embedding. This would enable us to set Python 3.2+ in package dependencies, build with whatever version we like and forget about it.

Please let me know which path you think is better to follow. I can work on static embedding next week.

**Comment by matiasinsaurralde** [comment]
We could drop the Cython generated code, it's mainly used to attach it to some of the gateway functions like `TykGetKey`, `TykSetKey`, etc., without writing this glue code in a more complex way.

**Comment by excieve** [comment]
@matiasinsaurralde Do I understand correctly that there are basically two ways to replace the Cython-generated module: using CFFI or a writing a C module by hand?

This glue code doesn't seem to do much indeed so it might be easy enough to actually just do the latter. I can look into it if you have other things to do.

**Comment by matiasinsaurralde** [comment]
That's right, it would be great if you can see how to do the static build, I can collaborate on tweaking the Tyk API calls.
Before Cython I've tried FFI and it worked fine but it was easy to have something like this: https://github.com/TykTechnologies/tyk/blob/master/coprocess/python/tyk/gateway.pyx

**Comment by nickReyn** [comment]
@buger to follow up

**Comment by buger** [comment]
@excieve #1338 is merged. Do you think we should close it, or re-use, or you create a separate issue for compiling with different python runtimes for different OS?

**Comment by altaurog** [comment]
This looks really promising, but I gather it isn’t ready for prime time?  I am running debian stretch with my own build of python3.6 installed and I would like to build tyk.  I plan to deploy it in a docker container, so of course I could just build it in a separate container as well, but latest python is preferred.  Is the [python readme](https://github.com/TykTechnologies/tyk/blob/master/coprocess/python/README.md) up-to-date?  It would be great to get some specific, up-to-date instructions.

**Comment by matiasinsaurralde** [comment]
@altaurog We're considering this for an upcoming release.
You may find the updated Python instructions [here](https://tyk.io/docs/customise-tyk/plugins/rich-plugins/python/).

**Comment by altaurog** [comment]
@matiasinsaurralde thanks.  It seems to have built master successfully.  I will test over the next few days.

**Comment by altaurog** [comment]
Still not able to get it running.  I reported the error I’m getting in #1804.

**Comment by matiasinsaurralde** [comment]
An update on this: I'm developing a package that tackles the issue described in this ticket by loading Python dynamically using `libdl`. Currently the code is available [here](https://github.com/matiasinsaurralde/dlpython), I've started to work on this a few weeks ago and I'm tweaking the binding generation mechanisms so that we avoid writing code for every CPython API call we require (wrapping a single call requires writing a few lines of C and Go code, the gateway uses around 20 different calls).
The impact on the gateway will be as follows:
- Most of the `cgo` code will be removed from the gateway repository.
- The binary will not be linked to `libpython3.4` anymore.
- We'll use the `python-config` tool to discover the Python library that's available in a given system, during runtime (have found that this is pretty standard, in any case I'm looking forward to extend the package tests to more distros to see how it behaves, at the moment I've tested different Ubuntu/Debian versions with different Python versions under each one, got positive results).

I'm pretty close to testing the gateway integration.
